pkg/csr/csr.go

```diff
@@ -0,0 +1,350 @@
+	certificates.UsageDigitalSignature,
+	certificates.UsageServerAuth,
+}
+
+const (
+	nodeGroup          = "system:nodes"
+	nodeUserName       = "system:node"
+	NodeUserNamePrefix = nodeUserName + ":"
+	systemPrefix       = "system:authenticated"
+)
+
+// Approver holds the information required to approve a node CSR
+type Approver struct {
+	// client is the cache client
+	client client.Client
+	// k8sclientset holds the kube client needed for updating CSR approval conditions
+	k8sclientset *kubernetes.Clientset
+	// csr holds the pointer to the CSR request
+	csr      *certificates.CertificateSigningRequest
+	log      logr.Logger
+	recorder record.EventRecorder
+	// namespace is the namespace in which CSR's are present
+	namespace string
+}
+
+// NewApprover returns a pointer to the Approver
+func NewApprover(client client.Client, clientSet *kubernetes.Clientset, csr *certificates.CertificateSigningRequest,
+	log logr.Logger, recorder record.EventRecorder, watchNamespace string) (*Approver, error) {
+	if client == nil || csr == nil || clientSet == nil {
+		return nil, errors.New("kubernetes client, clientSet or CSR should not be nil")
+	}
+	return &Approver{client,
+		clientSet,
+		csr,
+		log,
+		recorder,
+		watchNamespace}, nil
+}
+
+// Approve approves a CSR by updating its status conditions to true if it is a valid CSR
+func (a *Approver) Approve() error {
+	if a.k8sclientset == nil {
+		return errors.New("Kubernetes clientSet should not be nil")
+	}
+
+	if valid, err := a.validateCSRContents(); !valid {
+		// if the validation fails and error returned is nil, returns nil
+		return errors.Wrapf(err, "could not validate CSR %s contents for approval", a.csr.Name)
+	}
+
+	a.csr.Status.Conditions = append(a.csr.Status.Conditions, certificates.CertificateSigningRequestCondition{
+		Type:           certificates.CertificateApproved,
+		Status:         "True",
+		Message:        "Approved by WMCO",
+		LastUpdateTime: meta.Now(),
+		Reason:         "WMCOApproved",
+	})
+
+	if _, err := a.k8sclientset.CertificatesV1().CertificateSigningRequests().UpdateApproval(context.Background(),
+		a.csr.Name, a.csr, meta.UpdateOptions{}); err != nil {
+		return errors.Wrapf(err, "could not update conditions for CSR approval %s", a.csr.Name)
+	}
+	a.log.Info("CSR approved", "CSR", a.csr.Name)
+	return nil
+}
+
+// validateCSRContents returns true if the CSR request contents are valid.
+// If the CSR is not from a BYOH Windows instance, it returns false with no error.
+// If there is an error during validation, it returns false with the error.
+func (a *Approver) validateCSRContents() (bool, error) {
+	parsedCSR, err := ParseCSR(a.csr.Spec.Request)
+	if err != nil {
+		return false, errors.Wrapf(err, "error parsing CSR %s", a.csr.Name)
+	}
+
+	nodeName := strings.TrimPrefix(parsedCSR.Subject.CommonName, NodeUserNamePrefix)
+	if nodeName == "" {
+		return false, errors.Errorf("CSR subject name does not contain the required node user prefix %s",
+			a.csr.Name)
```

Comment:
This implies the node user prefix is `a.csr.Name`. Is that correct or should `%s` be moved after `CSR`?